pkg/networkservice/core/adapters/server_to_client.go

```diff
@@ -36,9 +38,27 @@ func NewServerToClient(server networkservice.NetworkServiceServer) networkservic
 }
 
 func (s *serverToClient) Request(ctx context.Context, in *networkservice.NetworkServiceRequest, _ ...grpc.CallOption) (*networkservice.Connection, error) {
-	return s.server.Request(ctx, in)
+	conn, err := s.server.Request(ctx, in)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		in = &networkservice.NetworkServiceRequest{}
+	}
+	in.Connection = conn
+	if !next.Done(ctx) {
```

Comment:
NOTE: I've added flag `done` for pkg next to support case when we have a chain of 
s -> adapted c - > s and for example client did broke the chain, then we should not call next server.